pkg/tfbridge/auto_aliasing.go

```diff
@@ -95,20 +95,75 @@ func (info *ProviderInfo) MustApplyAutoAliases() {
 // - Edit [SchemaInfo.MaxItemsOne]
 //
 // The goal is to always maximize backwards compatibility and reduce breaking changes for
-// the users of the Pulumi providers.
+// the users of the Pulumi providers. The basic functionality behind each action is
+// identical; ApplyAutoAliases keeps a record of which TF token maps to which Pulumi
+// token, which fields have MaxItemsOne (true or false), and what version the record is
+// from.
 //
-// Resource aliases help mask TF provider resource renames or changes in mapped tokens so
-// older programs continue to work.  See [ResourceInfo.RenameResourceWithAlias] and
-// [ResourceInfo.Aliases] for more details.
+// For example, this is the (abbreviated & modified) history for GCP's compute autoscalar:
```

Comment:
There's some missing context here that I think would be beneficial to go into first.

We are specifically solving the problem of backcompat with aliasing  👍  I would like an explanation/diagram of the specific mechanics of auto aliasing.

- where does this example come from/where do we keep this record?
- how is this history maintained and modified?

Reply:
I would like to avoid going too deep into that. The bridge has a system for maintaining these kinds of changes `github.com/pulumi/pulumi-terraform-bridge/unstable/metadata`. I mention that we use this system, but how that works is a separate thing, and not necessary for understanding `ApplyAutoAliases`.


Brief summary:

The data winds up in `bridge-metadata.json` (`."auto-aliasing"`) and its lifecycle is managed as part of `make tfgen`.

Retrieving from history:
https://github.com/pulumi/pulumi-terraform-bridge/blob/64ac968bae388b3e941ce688d447e65c197a48e7/pkg/tfbridge/auto_aliasing.go#L277-L290

Saving to history:
https://github.com/pulumi/pulumi-terraform-bridge/blob/64ac968bae388b3e941ce688d447e65c197a48e7/pkg/tfbridge/auto_aliasing.go#L266-L270